simpleclient_pushgateway/src/main/java/io/prometheus/client/exporter/PushGateway.java

```diff
@@ -50,28 +61,89 @@
  * }
  * </pre>
  * <p>
+ * Example usage with Basic authentication and SSL:
+ * <pre>
+ * {@code
+ *   void executeBatchJob() throws Exception {
+ *     ...
+ *     try {
+ *      ...
+ *     } finally {
+ *       durationTimer.setDuration();
+ *       CredentialsProvider creds = new BasicCredentialsProvider();
+ *       creds.setCredentials(
+ *          new AuthScope(new HttpHost("localhost")),
+ *          new UsernamePasswordCredentials("username", "password"));
+ *       CloseableHttpClient httpClient = HttpClients.custom().setDefaultCredentialsProvider(creds).build();
+ *       PushGateway pg = new PushGateway("https", "localhost:443", httpClient);
+ *       pg.pushAdd(registry, "my_batch_job");
+ *     }
+ *   }>
+ * }
+ * </pre>
+ * For information about authentication using the HttpClient see:
+ * <a href="https://hc.apache.org/httpcomponents-client-ga/tutorial/html/authentication.html">
+ * https://hc.apache.org/httpcomponents-client-ga/tutorial/html/authentication.html</a>
+ * <p>
+ * For preemptive-auth see:
+ * <a href="http://www.programcreek.com/java-api-examples/index.php?source_dir=jenkins-client-master/src/main/java/com/offbytwo/jenkins/client/JenkinsHttpClient.java">JenkingsHttpClient</a>
+ * <a href="http://www.programcreek.com/java-api-examples/index.php?source_dir=jenkins-client-master/src/main/java/com/offbytwo/jenkins/client/PreemptiveAuth.java">PreemptiveAuth</a>
+ * <p>
  * See <a href="https://github.com/prometheus/pushgateway">https://github.com/prometheus/pushgateway</a>
  */
 public class PushGateway {
 
-  private final String address;
-  private static final int SECONDS_PER_MILLISECOND = 1000;
+  private static final int TIMEOUT_IN_MILLISECONDS = 10000;
+
+  private final HttpClient httpClient;
+
+  private final String urlBase;
+
+  private BasicHttpContext localContext;
```

Comment:
Thats not 100% correct.
As mentioned in https://hc.apache.org/httpcomponents-core-ga/httpcore/apidocs/org/apache/http/protocol/BasicHttpContext.html the BasicHttpContext can be not threadsafe if the parent context isn't threadsafe. But we don't use a parent context so this basic context is threadsafe.
Do you agree?